File: Advanced Methods of Programming/Toy Language Interpreter/src/Model/ToyProgram.java
package Model;

import Model.Exceptions.ProgramCompletedException;
import Model.Statements.IStatement;

public class ToyProgram {
    private int id;
    private ProgramState state;
    private IStatement statement;

    public ToyProgram(IStatement statement){
        this.statement = statement;
        this.state = new ProgramState();
        state.getExeStack().push(statement);
        id = 1;
        state.setID(1);
    }

    private ToyProgram(ProgramState state){
        this.statement = state.getExeStack().peek();
        this.state = state;
    }

    public ToyProgram oneStep() throws ProgramCompletedException {
        if( this.completed() ){
            throw new ProgramCompletedException("All steps executed");
        }
        IStatement currentStatement = state.getExeStack().pop();
        ProgramState newState = currentStatement.exec(state);
        if (newState != null)
            return new ToyProgram(newState);
        return null;
    }

    public Boolean completed(){
        return state.getExeStack().empty();
    }

    public ProgramState getState(){ return state; }

    public String getOutput(){ return state.getOutput(); }

    public String toString(){ return statement.toString(); }

    public Integer getID() { return id; }

    public void setID(int id){
        this.id = id;
        this.state.setID(id);
    }
}

File: Advanced Methods of Programming/CSharp Toy Language Interpreter/ToyLanguageInterpreter/Src/Model/ToyProgram.cs
using System;
using ToyLanguageInterpreter.Src.Model.Exceptions;
using ToyLanguageInterpreter.Src.Model.Statements;

namespace ToyLanguageInterpreter.Src.Model
{
    class ToyProgram
    {
        public int id { get; set; }
        public ProgramState State { get; set; }
        public IStatement Statement { get; set; }
        public string Output
        {
            get { return State.Output; }
        }

        public ToyProgram(IStatement statement)
        {
            this.Statement = statement;
            this.State = new ProgramState();
            State.ExeStack.Push(statement);
            id = 1;
        }

        private ToyProgram(ProgramState state)
        {
            this.Statement = state.ExeStack.Peek();
            this.State = state;
        }

        public ToyProgram OneStep()
        {
            if(this.Completed())
            {
                throw new ProgramCompletedException("All steps executed");
            }

            IStatement currentStatement = State.ExeStack.Pop();
            ProgramState newState = currentStatement.Exec(State);

            if (newState != null)
                return new ToyProgram(newState);

            return null;
        }

        public bool Completed()
        {
            return State.ExeStack.Empty();
        }

        public override string ToString()
        {
            return Statement.ToString();
        }

    }
}
